src/unix/lwt_io.ml

```diff
@@ -1454,6 +1454,49 @@ let with_file ?buffer ?flags ?perm ~mode filename f =
     (fun () -> f ic)
     (fun () -> close ic)
 
+let prng = lazy(Random.State.make_self_init ())
+
+let temp_file_name temp_dir prefix =
+  let rnd = (Random.State.bits (Lazy.force prng)) land 0xFFFFFF in
+  Filename.concat temp_dir (Printf.sprintf "%s%06x" prefix rnd)
+
+let open_temp_file:
+  ?buffer : Lwt_bytes.t ->
+  ?flags : Unix.open_flag list ->
+  ?perm : Unix.file_perm ->
+  ?temp_dir : string ->
+  ?prefix : string ->
+  unit -> (string * output_channel) Lwt.t
+  = fun ?buffer ?flags ?perm ?temp_dir ?prefix->
+    let flags = match flags with
+    | None -> Some [Unix.O_WRONLY; Unix.O_CREAT; Unix.O_EXCL; Unix.O_CLOEXEC]
+    | Some l -> Some l in
+    let dir = match temp_dir with
+    | None -> Filename.get_temp_dir_name ()
+    | Some dirname -> dirname in
+    let prefix = match prefix with
+    | None -> "lwt_io_temp_file_"
+    | Some prefix -> prefix in
+    let rec helper n =
+      let fname = temp_file_name dir prefix in
+      Lwt.catch
+        (fun () -> open_file ?buffer:buffer ?flags:flags ?perm:perm ~mode:Output fname
+          >>= fun chan -> Lwt.return (fname, chan))
+        (fun exn ->
+           match exn with
+           | Unix.Unix_error _ when n < 1000 -> helper (n + 1)
+           | _ -> raise exn)
+    in
+    fun () -> helper 0
```

Comment:
Ah, this is still an example of doing side effects before full application (in this case, I believe a coeffect, as applying `open_temp_file` to all its labeled arguments but *not* `()` will cause it to read `Filename.get_temp_dir_name ()` if `?temp_dir:None`, which "fixes" the temporary directory, even if it is changed before applying to `()`, or between multiple applications to `()`). Sorry for missing it, I'll just tweak it in `master`.